tests/contrib/gp/test_models.py

```diff
@@ -71,29 +98,195 @@ def test_model_forward(model_class, X, y, kernel, likelihood):
+    loc_diff = loc - loc[..., :1].expand(y.shape[:-1] + (10,))
+    assert_equal(loc_diff.norm().item(), 0)
+    cov_diff = cov - cov[..., :1, :1].expand(y.shape[:-1] + (10, 10))
+    assert_equal(cov_diff.norm().item(), 0)
+
+    # test noise kernel forward: kernel = WhiteNoise
+    gp.kernel = WhiteNoise(input_dim=3, variance=torch.tensor(10))
+    Xnew = torch.tensor([[2, 3, 1]])
+    loc, cov = gp(X, full_cov=True)
+    assert_equal(loc.norm().item(), 0)
+    assert_equal(cov, torch.eye(cov.shape[-1]).expand(cov.shape) * 10)
 
 
 @pytest.mark.parametrize("model_class, X, y, kernel, likelihood", TEST_CASES, ids=TEST_IDS)
+def test_forward_with_empty_latent_shape(model_class, X, y, kernel, likelihood):
+    # regression models don't use latent_shape, no need for test
+    if model_class is GPRegression or model_class is SparseGPRegression:
+        return
+    elif model_class is VariationalGP:
+        gp = model_class(X, y, kernel, likelihood, latent_shape=torch.Size([]))
+    else:  # model_class is SparseVariationalGP
+        gp = model_class(X, y, kernel, X, likelihood, latent_shape=torch.Size([]))
+
+    # test shape
+    Xnew = torch.tensor([[2, 3, 1]])
+    loc0, cov0 = gp(Xnew, full_cov=True)
+    loc1, var1 = gp(Xnew, full_cov=False)
+    assert loc0.shape[-1] == Xnew.shape[0]
+    assert cov0.shape[-1] == cov0.shape[-2]
+    assert cov0.shape[-1] == Xnew.shape[0]
+    # test latent shape
+    assert loc0.shape[:-1] == torch.Size([])
+    assert cov0.shape[:-2] == torch.Size([])
+    assert_equal(loc0, loc1)
+    assert_equal(cov0.diag(), var1)
+
+
+@pytest.mark.parametrize("model_class, X, y, kernel, likelihood", TEST_CASES, ids=TEST_IDS)
+@pytest.mark.init(rng_seed=0)
+@pytest.mark.disable_validation()
 def test_inference(model_class, X, y, kernel, likelihood):
+    # skip variational GP models because variance/lengthscale highly
+    # depend on variational parameters
+    if model_class is VariationalGP or model_class is SparseVariationalGP:
+        return
+    elif model_class is GPRegression:
+        gp = model_class(X, y, RBF(input_dim=3), likelihood)
+    else:  # model_class is SparseGPRegression
+        gp = model_class(X, y, RBF(input_dim=3), X, likelihood)
+        # fix inducing points because variance/lengthscale highly depend on it
+        gp.fix_param("Xu")
+
+    generator = dist.MultivariateNormal(torch.zeros(X.shape[0]), kernel(X))
+    target_y = generator(sample_shape=torch.Size([1000])).detach()
+    gp.set_data(X, target_y)
+
+    gp.optimize(optim.Adam({"lr": 0.01}), num_steps=1000)
+
+    variance = gp.kernel.get_param("variance")
+    lengthscale = gp.kernel.get_param("lengthscale")
+    target_variance = kernel.get_param("variance")
+    target_lengthscale = kernel.get_param("lengthscale")
+    assert_equal(variance, target_variance, prec=0.2)
+    assert_equal(lengthscale, target_lengthscale, prec=0.2)
+
+
+@pytest.mark.init(rng_seed=0)
+@pytest.mark.disable_validation()
```

Comment:
Can you run with validation for one step, then disable and continue for 1000 steps?